Add K8s runtime

mruoss opened the pull request:
Alright, @josevalim asked me if I wanted to contribute a K8s runtime so here it comes. Took me ages to build a UI with a few forms on it!!! 🤣 

## TODOs

- [x] Code Review suggestions
- [x] Open discussion on abort functionality and timeout
- [x] UI tests
- [x] integration tests
- [x] docs and specs
- [x] open suggestion on `start_runtime.exs`
- [x] Changelog entry


## A few notes

### Connection and Authentication

When working with Kubernetes you usually don't ever see connection and authentication details. Instead, you use the helper of your cloud provider which adds all infos into your user's Kubernetes configuration YAML file - on unix systems that's on `~/.kube/config`. I therefore decided to rely on an ENV variable pointing to such a Kubeconfig file instead of prompting for credentials. If users want an ad-hoc way of connecting to a cluster, I can add this in a later PR. But I'd rather let people use it and gather some feedback.

### Specs

Kubernetes Pod specs contains soo many fields. I'm only allowing to configure the ones I think are most important especially and also for running on GPU nodes (node selector, tolerations, resources, persistent volume). For FLAME I support a BYO Pod manifest. This might be something to add later so that people can paste their YAML instead of filling the form. We would than only override the image and add required ENV vars.

### UI

Yeah... please somebody make that form more beautiful! 😜 No seriously, I think it's ok for now but I'm sure somebody could do a better job.

![Screenshot 2024-08-19 at 21 55 10](https://github.com/user-attachments/assets/790b6297-69e4-4e8c-a56a-c44b8ee6bd43)

### Ready for Review

I'm still gonna run some tests (in-cluster and error handling for missing permissions and other non-happy cases) and maybe commit some fixes but feel free to review what's there.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/livebook-dev/livebook?pullRequest=2756) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/livebook-dev/livebook?pullRequest=2756) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: mruoss<br/>:x: Michael Ruoss<br/><hr/>**Michael Ruoss** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/livebook-dev/livebook?pullRequest=2756) it.</sub>

josevalim commented:
Thank you @mruoss! The PR looks great, I think the biggest questions are exactly the ones you wrote in the issue. Authentication and specifications.

About authentication, the most likely scenario is that I am using my desktop and I want to connect to a k8s cluster my company runs. Would this also be done via `~/.kube/config`?

About the specs, I think replicating it in the UI is just too much hassle. I can think of two options:

1. Templates? Is it possible to specify templates within k8s? If so, maybe we could simply list all deployments and we use the template in that deployment? Then we can provide some documentation on how to create a default deployment.

2. Another option is to just have a text area with the yaml . We can include a default yaml and we just use that. Later we can even make it use the codemirror editor. I think this may provide the better experience. We can simplify all of the code, users can copy and paste, and make it as complex and as simple they need it to be.

mruoss commented:
> About authentication, the most likely scenario is that I am using my desktop and I want to connect to a k8s cluster my company runs. Would this also be done via `~/.kube/config`?

If the company cluster's API Server is reachable, then it is technically possible, yes. But there are some caveats as I already mentioned earlier. Kubernetes supports several mechanisms of authentication. Some of these mechanism (`exec`) require a shell for a user that is logged in to the cloud provider (e.g. GCP). These auth mechanisms won't work for let's say livebook dekstop. They only work if you run livebook from the CLI as the user that's logged in to the cloud provider.

There's a workaround for this use case, though. If the user has enough permissions, they can create a K8s Service Account (SA) or use an existing SA and make sure it has all the required permissions (Role and RoleBinding). They can then [Create an API Token for the SA](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount) and create new cluster, user and context entries in `~/.kube/config` or create a new and dedicated kubeconfig file. This is not very difficult but you have to know what you're doing. I could create a mix task that does all of this. But then again, when Livebook is run as Desktop version, there's no mix tasks to the rescue.

To come back to your initial question: I _think_ most/all auth mechanism are supported by `Kubereq` via Kubeconfig file. But creating UIs for all of them (API Token, Client Cert auth, ...) would be a lot of work. That's the reason I decided to go with the file. It's just the "creating the file" that's the difficult part. 

Hope I could make things clear.


> About the specs, I think replicating it in the UI is just too much hassle. I can think of two options:

Are you proposing to remove the form completely and _only_ offer a BYO manifest version? Or leave the simple form and offer a more sophisticated alternative?

> 1. Templates? Is it possible to specify templates within k8s? If so, maybe we could simply list all deployments and we use the template in that deployment? Then we can provide some documentation on how to create a default deployment.

It is a nice thought. "Deployment" is the wrong term here, because a Deployment actually schedules pods to be run. But I think there is a `PodTemplate` resource we could use. Users would then have to create a PodTemplate and apply it to the cluster prior to creating the runtime. I'll dig a bit.

> 2. Another option is to just have a text area with the yaml . We can include a default yaml and we just use that. Later we can even make it use the codemirror editor. I think this may provide the better experience. We can simplify all of the code, users can copy and paste, and make it as complex and as simple they need it to be.

Yup, that's the thought I had as well. Both options would require users to also care about PVCs. Mixing UI/Form elements with BYO is also going to be too bit of a hassle...

Let me give the options a bit more thought....

josevalim commented:
> To come back to your initial question: I think most/all auth mechanism are supported by Kubereq via Kubeconfig file. But creating UIs for all of them (API Token, Client Cert auth, ...) would be a lot of work. That's the reason I decided to go with the file. It's just the "creating the file" that's the difficult part.

To be clear, I actually like the file approach. The less we have to do, the better. :) I was just wondering how users would get that file in non-conventional cases.

> Are you proposing to remove the form completely and only offer a BYO manifest version? Or leave the simple form and offer a more sophisticated alternative?

Correct. I would get rid of the form altogether. Almost everyone using k8s will be familiar with the yaml config. We can support either YAML, PodTemplates, or both. For PVC, we could still manage it, and either dynamically modify the yaml before applying it OR just change the default YAML we will show based on your PVC config.

mruoss commented:
> I would get rid of the form altogether. 

You've got to be kidding me! There is blood in this form! 😂 Just kidding. It will be much easier this way to maintain. I'll try to find a good way to support YAMLs.

josevalim commented:
> It will be much easier this way to maintain. I'll try to find a good way to support YAMLs.

Feel free to start with just the textarea and we can do the rest. :)

mruoss commented:
Okay, it's just a plain monospace textarea now. Using the codemirror editor instead would be amazing of course! But I didn't look into that yet.
I can definitely also add a "load PodTemplate from cluster" functionality. Although I'm not sure this really adds value for anybody.

josevalim commented on lib/livebook/k8s/kubeconfig.ex:
```diff
@@ -0,0 +1,7 @@
+defmodule Livebook.K8s.Kubeconfig do
```
I think we should roll with the default kubeconfig. I don't see a reason for supporting LIVEBOOK_KUBECONFIG because:

1. setting KUBECONFIG or LIVEBOOK_KUBECONFIG for the desktop or the CLI is the same amount of work
2. setting KUBECONFIG has the benefit that, if you want to use Kubereq inside the runtime, it will just work too

Plus, we probably want to support reading from `~/.kube/config` anyway,  no? This means we can also remove `:pluggable` from the .formatter.exs file :)

mruoss replied:
> This means we can also remove :pluggable from the .formatter.exs file 

Because we can use `https://github.com/mruoss/kubereq/blob/main/lib/kubereq/kubeconfig/default.ex` you mean?

josevalim replied:
Correct!

josevalim commented on rel/server/overlays/bin/start_runtime.exs:
```diff
@@ -6,10 +6,19 @@ File.cd!(System.user_home!())
+    System.get_env("POD_IP") ->
+      # This is the only K8s-specific part of starting Livebook as runtime
+      hostname = System.fetch_env!("POD_IP")
+      :"#{node_base}@#{hostname}"
```
```suggestion
      :"#{node_base}@#{hostname}"

    true ->
      raise "expected either POD_IP (for k8s) or FLY_APP_NAME (for Fly.io) to be set"
```

mruoss replied:
oh yes, forgot this one.

josevalim commented on lib/livebook/k8s/key_value.ex:
```diff
@@ -0,0 +1,20 @@
+defmodule Livebook.K8s.KeyValue do
```
I don't think this is used anymore!

mruoss replied:
true.

josevalim commented on lib/livebook/k8s/pod.ex:
```diff
@@ -0,0 +1,167 @@
+  @home_pvc_volume_name "lb-home-folder"
+  @default_pod_template """
+  apiVersion: v1
+  kind: PodTemplate
```
Sorry, tiny nitpick, but I don't think this needs to be a template, right? it can just be directly the pod configuration. It seems all we do with the template is to extract the template key anyway, so we might as well get rid of the indirection?

mruoss replied:
Since this particular resource is never stored on the cluster, it would technically be okay to make it a Pod. 
I gave cluster-stored PodTemplate resources some more thought though. Maybe there is actually a use case for supporting them. E.g. if you want to have a set of different templates for different GPUs (have a template for a100 nodes, one for T4, etc.), you could create these resources on the cluster and then just select the correct one according to your needs. 
Or you think that's too far-fetched?

mruoss replied:
I have updated the code. It now supports `Pod` and/or `PodTemplate` manifests. I have made the default config a `Pod` though as you are right, the extra metadata didn't really add any value.

josevalim replied:
> It now supports Pod and/or PodTemplate manifests.

Honestly, let's support only Pod, unless someone really needs template. Because otherwise you will also have to support PodTemplates on `flame_k8s_backend` as well. Fewer options for now is better!

josevalim commented on lib/livebook_web/components/core_components.ex:
```diff
@@ -930,7 +930,7 @@ defmodule LivebookWeb.CoreComponents do
   attr :class, :string, default: nil
 
-  attr :rest, :global, include: ~w(href patch navigate download name)
+  attr :rest, :global, include: ~w(href patch navigate download name value)
```
Yes please!

mruoss commented:
Okay. I have one last thing on my mind/list: I think I need to make the timeout (how long to wait for the pod to get scheduled and start running) configurable. In some cases, this can take minues (e.g. downloading the docker image already can take a minute). 
Also the default probably has to be higher than 15s.

josevalim commented:
Perhaps we hust make the timeout arbitrarily high and we allow the user to press the stop button if they want to give up?

mruoss commented:
Or that... And find a way to show event stream (e.g. "downloading image...)

mruoss commented:
I've pushed a few commits today which make the whole feature more robust (checks for all required permissions, error messages etc). Now I am okay with the code all-in all with the following exception:

I have put the timeout to 5 minutes now but can even make it higher. However, I didn't really figure out a good way to abort the connection. What I have implemented now is that the runtime updates (`send_update/3`) the runtime LV component and assigns the `pod_name`. If that variable is non-nil in the runtime component, it renders an "abort connection" button which snipes the pod. See https://github.com/livebook-dev/livebook/pull/2756/commits/951bfd602592e1f5d6951212727e44f65e7e1fed.

I don't like the solution too much because of the messaging back and also because the user gets an error message saying that the Pod was deleted... So if there's a better way to abort the connection, let me know.

jonatanklosko commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,566 @@
+             create_pod(req, config, runtime_data, cluster_data.remote_port)
+           end),
+         _ <-
+           Phoenix.LiveView.send_update(
+             runtime.lv_pid,
+             LivebookWeb.SessionLive.K8sRuntimeComponent,
+             id: "runtime-config-k8s",
+             pod_name: pod_name
+           ),
```
We should not communicate with the LV directly. I generally wouldn't use `send_update` outside LV code, because this introduces brittle coupling.

For a more practical reason, when the user refreshes the page this would no longer work (or for another user looking at the UI).

---

Let's remove this part. I need to think more how to handle cancellation from the UI. Technically, I think it would be fine to just kill the process, because the resources are designed to autoterminate (e.g. once the pod comes up it would quickly terminate itself).

mruoss replied:
Yeah I've also noticed that when reconnecting using the button in the left sidebar, this doesn't really work. Left it in for the discussion. 

So you're proposing to kill the runtime process? Do we get its pid in the LV? And I guess I add a button to the sidebar as well?

jonatanklosko replied:
Let's skip this for now. I think it would also be useful for the Fly runtime, so I will have a look at that and then we can do the same here.

I think killing the process is the way to go, I just need to fit it into the runtime API. If we want to eagerly delete the pod on abort, then we can do still do that. We would spawn a watcher process that monitors the caller. Once the pod is created we send info to the watcher. When the caller terminates unexpectedly, the watcher can do eager cleanup.

mruoss replied:
You want to skip the whole abort functionality for now? The problem is: It is quite easy to create a manifest for a pod that can't be scheduled by the Scheduler. Let's say I want to connect to my cluster but I'd like the node to have a GPU. So I add `nvidia.com/gpu: "1"` to my resources but often those nodes are tainted and I forgot to put the toleration for that taint. The pod cannot be scheduled and we have to wait the full timeout (5/10/15 minutes?). 

I can skip it in this PR. Just saying I think it should be addressed as this can lead to frustration.

jonatanklosko replied:
> You want to skip the whole abort functionality for now?

Yes, but my plan is to add abort for Fly runtime soon and then we mirror it in this PR or add afterwards, but definitely before the next release :)

jonatanklosko replied:
@mruoss see `fly.ex` and `fly_component.ex` in #2776, let's do the same here. If there are any issues, let me know!

mruoss replied:
Nice! Will do.

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,932 @@
+  end
+
+  @impl true
+  def handle_event("validate_kubeconfig", _params, socket) do
+    {:noreply, socket}
+  end
+
+  def handle_event("save_kubeconfig", _params, socket) do
+    {:noreply, socket}
+  end
+
```
I think these are no longer used :)

```suggestion
```

jonatanklosko commented on lib/livebook/k8s/pod.ex:
```diff
@@ -0,0 +1,170 @@
+    end
+  end
+
+  defguardp is_empty(value) when is_nil(value) or value == "" or value == []
```
```suggestion
  defguardp is_empty(value) when value in [nil, "", []]
```

jonatanklosko commented on lib/livebook/k8s/pod.ex:
```diff
@@ -0,0 +1,170 @@
+            memory: 1Gi
+  """
+
+  defmacrop access_main_container() do
```
This can be a regular function :)

mruoss replied:
oh true.

jonatanklosko commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,566 @@
+defimpl Livebook.Runtime, for: Livebook.Runtime.K8s do
+  alias Livebook.Runtime.ErlDist.RuntimeServer
+
+  def describe(_) do
+    [
+      {"Type", "K8s Pod"}
+    ]
+  end
```
```suggestion
  def describe(runtime) do
    [{"Type", "K8s Pod"}] ++
      if runtime.node do
        [{"Node name", Atom.to_string(runtime.node)}]
      else
        []
      end
  end
```

jonatanklosko reviewed (commented):
Thanks for the PR @mruoss! I left some comments inline. A few additional notes:

1. It would be great if we could add an integration test (against a local cluster) for the runtime itself, similar to `test/livebook/runtime/fly_test.exs`. This way we have an easy way to assert all the connection and communication, which is the most important. It would only run when `TEST_KUBECONFIG` or similar is set. I would appreciate if you could also add a comment with minimal steps necessary for a matching local k8s setup.

2. We should have a test for the UI. For this one we want to stub all the requests. Similar test for Fly [here](https://github.com/livebook-dev/livebook/blob/c8e90073e95a1bef42758515350b9ab684ea3fdc/test/livebook_web/live/session_live_test.exs#L951).

3. All public functions in the codebase should have docs and specs. It may not feel necessary in all cases, but I think sticking to this rule makes the codebase more accessible.

4. We can definitely share some code with the Fly runtime, but I will do that separately!

jonatanklosko commented on rel/server/overlays/bin/start_runtime.exs:
```diff
@@ -6,10 +6,22 @@ File.cd!(System.user_home!())
+
+    true ->
+      raise "expected either POD_IP (for k8s) or FLY_APP_NAME (for Fly.io) to be set"
+  end
```
We can make it more generic:

```elixir
node =
  node_template
  |> String.replace(~r/\{\w+?\}/, fn substitute ->
    env = binary_slice(substitute, 1..-2//1)
    System.fetch_env!(env)
  end)
  |> String.to_atom()
```

The above replaces any `{ENV_NAME}` with the env var value.

Then, instead of passing `node_base`, we pass `node_template` as `"#{node_base}@{FLY_MACHINE_ID}.vm.{FLY_APP_NAME}.internal"` and `"#{node_base}@{POD_IP}"` respectively.

@josevalim wdyt?

josevalim replied:
Last time I was arguing to make it more generic and you wanted to keep it simple, and I think this time it will be the opposite. I believe the env vars are fine as and with less indirection, I would be glad to ship it as is. :)

jonatanklosko replied:
We discussed offline, @mruoss let's keep it as is :)

mruoss commented:
> 1. It would be great if we could add an integration test (against a local cluster) for the runtime itself, similar to `test/livebook/runtime/fly_test.exs`. This way we have an easy way to assert all the connection and communication, which is the most important. It would only run when `TEST_KUBECONFIG` or similar is set. I would appreciate if you could also add a comment with minimal steps necessary for a matching local k8s setup.

Oh yeah, absolutely agree. For `flame_k8s_backend` I have implemented the `setup_all/N` callback to [check for a local Kind cluster and create one if it is not there etc.](https://github.com/mruoss/flame_k8s_backend/blob/main/test/integration/integration_test.exs#L4) It is super handy for me because I just need to run `mix test --include integration` and the env is setup for me. Not sure though if you want to go that far here?

I can document K8s dev setup using markdown or by implementing a mix task - up to you.

> 2. We should have a test for the UI. For this one we want to stub all the requests. Similar test for Fly [here](https://github.com/livebook-dev/livebook/blob/c8e90073e95a1bef42758515350b9ab684ea3fdc/test/livebook_web/live/session_live_test.exs#L951).

Sure will do. Just wanted to get to an agreement on the concept / basic implementation first. 

> 3. All public functions in the codebase should have docs and specs. It may not feel necessary in all cases, but I think sticking to this rule makes the codebase more accessible.

Oh yeah no problem, on my libs I even go further than this.

> 4. We can definitely share some code with the Fly runtime, but I will do that separately!

👍

mruoss commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,932 @@
+  defp set_namespace(socket, ns) do
+    reqs = socket.assigns.reqs
+
+    with :ok <- Auth.can_i?(reqs, verb: "get", version: "v1", resource: "pods", namespace: ns),
```
Yes, you're right. can be async and run concurrently.

mruoss replied:
pushed.

mruoss commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,566 @@
+      runtime_pid = self()
+
+      event_watcher_pid =
+        spawn(fn ->
```
The reason I didn't link the processes (correct me if I'm wrong): The current user might not have permissions to watch pods which is okay because this basically just adds more info. But if he doesn't have the permissions, the spawned process crashes and I didn't want it to pull the parent down as well. 

But now that I think about it, I can also just make the watcher a bit more fault tolerant...

jonatanklosko replied:
Yeah, instead of crashing in the watcher, we can match on error and just let the process finish normally, so it doesn't bring the caller down :)

jonatanklosko commented:
> For flame_k8s_backend I have implemented the setup_all/N callback to [check for a local Kind cluster and create one if it is not there etc.](https://github.com/mruoss/flame_k8s_backend/blob/main/test/integration/integration_test.exs#L4)

Sounds great!

mruoss commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,566 @@
+                   "status"
+                 ]) == ["True"]
+             end,
+             # 24 hours (try "forever" and let peopel stop manually)
+             86_400_000
```
I'm having a hard time setting this to a fix number. Sometimes it can take a few minutes for sure. First the scheduler has to find a node that suits the needs. Maybe a new node needs to be provisioned first. Once the pod can be scheduled, the image has to be downloaded. 

Also see the comment on the abort functionality.

Will set to 10 minutes for now.

jonatanklosko replied:
Any reasonable upper bound work for me, can be 30 minutes. I just wouldn't make it hours.

mruoss commented:
@jonatanklosko I've added an integration test now. I've also added it to the CI because I think this really helps to catch regression. However, it does slow down the pipeline quite a bit (40-50s to install Kind and the cluster plus >1min to run the test as in the `setup_all` I have to build and load the docker image to the cluster. WDYT?

jonatanklosko commented:
@mruoss honestly I think it's fine to skip it on CI. The tested code is rather encapsulated, so it should be evident when we want to run these tests. The Fly runtime tests are already opt-in, because they require a token. This is definitely a trade-off, but I think slowing the CI feedback by 1min+ is not worth it :)

mruoss commented:
OK I will remove it from CI again. Can always put it back later. Also possible in a different pipeline that runs under different conditions.

josevalim commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,559 @@
+      |> Pod.pod_from_template()
+      |> Pod.add_env_vars([
+        %{"name" => "LIVEBOOK_RUNTIME", "value" => runtime_data},
+        %{"name" => "ERL_AFLAGS", "value" => "-proto_dist inet6_tcp"},
```
Is it advised to use ipv6 with k8s?

mruoss replied:
ah no... probably copy-paste. I will remove

josevalim commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,559 @@
+             create_pod(req, config, runtime_data, cluster_data.remote_port)
+           end),
+         {:ok, pod_ip} <-
+           with_pod_events(caller, "Waiting for pod to get ready", req, namespace, pod_name, fn ->
```
Lowercase for consistency and making it shorter:

```suggestion
           with_pod_events(caller, "waiting for pod", req, namespace, pod_name, fn ->
```

jonatanklosko commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,558 @@
+          case watch_result do
+            {:ok, stream} ->
+              Enum.each(stream, fn event ->
+                Logger.debug(~s'[K8s runtime] Pod event: "#{event["object"]["message"]}"')
+                send(caller, {:runtime_connect_info, runtime_pid, event["object"]["message"]})
```
Let's downcase the first character, so it's consistent with other info strings we send

```suggestion
                message = Livebook.Utils.downcase_first(event["object"]["message"])
                Logger.debug(~s'[K8s runtime] Pod event: "#{message}"')
                send(caller, {:runtime_connect_info, runtime_pid, message})
```

mruoss replied:
ahm... sure... :)

jonatanklosko commented on test/livebook/runtime/k8s_test.exs:
```diff
@@ -0,0 +1,141 @@
+    {clusters_out, exit_code} = System.cmd("kind", ~w(get clusters))
+    assert 0 == exit_code, "kind is not installed. Please install kind."
+
+    if not (clusters_out
+            |> String.split("\n", trim: true)
+            |> Enum.member?(@cluster_name)) do
```
```suggestion
    clusters = String.split(clusters_out, "\n", trim: true)

    if @cluster_name not in clusters do
```

jonatanklosko commented on test/livebook/runtime/k8s_test.exs:
```diff
@@ -0,0 +1,141 @@
+    if not (clusters_out
+            |> String.split("\n", trim: true)
+            |> Enum.member?(@cluster_name)) do
+      exit_code = Mix.Shell.IO.cmd("kind create cluster --name #{@cluster_name}", quiet: true)
```
Let's use `System.cmd` in all cases. We could use `Mix.shell().cmd`, but I don't think it's necessary here.

```suggestion
      {_, exit_code} = System.cmd("kind create cluster --name #{@cluster_name}")
```

mruoss replied:
```
  0) Livebook.Runtime.K8sTest: failure on setup_all callback, all tests have been invalidated
     ** (UndefinedFunctionError) function Mix.Shell.cmd/1 is undefined or private. Did you mean:

           * cmd/2
           * cmd/3
```

mruoss replied:
System.cmd had a totally different API. But I can rewrite it, no big deal... Any reason for this?

mruoss replied:
But also, ethere is no `:quiet` option which makes the test super verbose (prints all stdout from docker build etc)!

josevalim replied:
Do `into: ""`, and it will output into the string, which you can ignore!

mruoss replied:
mmh... according to the docs, `into: ""` is the default. And it still prints to stdout. You mean sth like this, right?

```ex
    System.cmd(
      "docker",
      ~w(buildx build --build-arg BASE_IMAGE=hexpm/elixir:#{bindings[:elixir]}-erlang-#{bindings[:otp]}-ubuntu-#{bindings[:ubuntu]} --build-arg VARIANT=default --load -t ghcr.io/livebook-dev/livebook:nightly .),
      into: ""
    )

```

josevalim replied:
ah, so stderr_to_stdout: true?

mruoss replied:
that just redirects stderr to stdout, no? But let me try

josevalim replied:
yeah, but stdout goes into the string, so we should be good !

mruoss replied:
oh so this verbose output was all stderr?? no way...

mruoss replied:
anyway, it does work. I really don't understand why. But I've pushed the changes.

jonatanklosko commented on test/livebook/runtime/k8s_test.exs:
```diff
@@ -0,0 +1,141 @@
+
+    assert [_] = list_pods(req)
+
+    # Verify that we can actually evaluate code on the Fly machine
```
```suggestion
    # Verify that we can actually evaluate code on the K8s pod
```

jonatanklosko commented on test/test_helper.exs:
```diff
@@ -72,8 +73,9 @@ teams_exclude =
   end
 
 fly_exclude = if System.get_env("TEST_FLY_API_TOKEN"), do: [], else: [:fly]
+k8s_exclude = if is_nil(System.find_executable("kind")), do: [:k8s], else: []
```
```suggestion
k8s_exclude = if System.find_executable("kind"), do: [], else: [:k8s]
```

jonatanklosko commented on lib/livebook/k8s/pod.ex:
```diff
@@ -0,0 +1,168 @@
+            memory: 1Gi
+  """
+
+  def access_main_container() do
```
```suggestion
  defp access_main_container() do
```

mruoss replied:
oops!

mruoss commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,925 @@
+  end
+
+  def set_pod_template(socket, pod_template_yaml) do
+    with {:ok, pod_template} <- YamlElixir.read_from_string(pod_template_yaml),
```
true.

mruoss commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,566 @@
+    %{node_base: "remote_runtime_#{local_port}", remote_port: 44444, local_port: local_port}
+  end
+
+  defp k8s_forward_port(nil, _, _, _, _), do: :ok
```
> This is when running Livebook itself on K8s, right? 

Yes exactly. kubeconfig is `nil` if we're running in a Kubernetes cluster:

```ex
    kubeconfig =
      if System.get_env("KUBERNETES_SERVICE_HOST"),
        do: nil,
        else: System.get_env("KUBECONFIG") || Path.join(System.user_home!(), ".kube/config")

    cluster_data = get_cluster_data(kubeconfig)
```

But also `get_cluster_data/1` is different. When running in a cluster, we set `node_base` to `"k8s_runtime"` instead of `"remote_runtime_#{local_port}"`. Later in the continue we then set `child_node = :"#{cluster_data.node_base}@#{pod_ip}"`. 
So if we're running inside a cluster, we're correctly targeting the pod's IP address while when running locally, that part doesn't matter and loopback is targeted.

mruoss commented:
I think it's only UI tests left now. Have to find a good way to mock the kubeconf setup first. Mocking the req should then be straight forward...

mruoss commented:
> @mruoss honestly I think it's fine to skip it on CI. 

@jonatanklosko It looks like `kind` is installed on those GH runners which makes the test run automatically. Shall I add `--exclude k8s` to `mix test` in the CI job?

jonatanklosko commented on .github/workflows/test.yml:
```diff
@@ -41,7 +41,7 @@ jobs:
         run: mix compile --warnings-as-errors
       - name: Run tests
-        run: mix test
+        run: mix test --exclude k8s
```
In that case, I would always exclude it, so it would be run only when `mix test --include k8s`.

jonatanklosko commented on lib/livebook/k8s/pod.ex:
```diff
@@ -0,0 +1,203 @@
+            memory: 1Gi
+          requests:
+            cpu: "1"
+            memory: 1Gi
```
```suggestion
            memory: 1Gi\
```

jonatanklosko commented:
@mruoss I did small adjustments to the form. I grouped the image picker and template under the same Pod section and moved storage to the bottom, since it's more of an addition to the template. Let me know if this makes sense :)

mruoss commented on lib/livebook/k8s/pod.ex:
```diff
@@ -0,0 +1,203 @@
+    generateName: livebook-runtime-
+  spec:
+    containers:
+      - image: ghcr.io/livebook-dev/livebook:nightly
```
Oh you're right. This was in the back of my head and I lost track. The current component implementation allows setting a custom image but warns that it should base off the current image:

![Screenshot 2024-09-10 at 12 57 06](https://github.com/user-attachments/assets/ec98bb55-0f09-4fff-8084-610529540418)

However, this is not implemented in the runtime and the selected livebook.dev image is used instead. I'd say I remove this functionality from the component for now. Can always add it later if people are baking their own, customized livebook images.

jonatanklosko replied:
Sounds good to me!

mruoss commented:
@jonatanklosko @josevalim I've added integration and UI test and feel pretty confident now regarding regression. So I think this is ready for a final review. I guess the changelog entry will be created on the 1.14 branch?

Sorry for taking so long but I didn't have that much time to work on it. That being said, I really really enjoyed working on this and refreshing my LiveView knowledge a little. There's been a lot of progress since the last time I've worked on a productive LV project and oh man... the whole framework feels so nice! Even writing tests, mocking calls etc. It all feels clear and easy and I had so much fun.

jonatanklosko commented on test/livebook/runtime/k8s_test.exs:
```diff
@@ -0,0 +1,142 @@
+      namespace: "default",
+      home_pvc: nil,
+      docker_tag: "nightly",
+      pod_template: %{template: @default_pod_template}
```
```suggestion
      pod_template: @default_pod_template
```

The map is only used in the UI assigns, no?

mruoss replied:
oh true - I forgot that the integration tests are excluded now.

jonatanklosko commented on lib/livebook/k8s/auth.ex:
```diff
@@ -0,0 +1,63 @@
+  def batch_check(req, resource_attribute_list) do
+    resource_attribute_list
+    |> Enum.map(&Task.async(fn -> can_i?(req, &1) end))
+    # 30 second timeout
+    |> Task.await_many(30_000)
```
Each request itself has a timeout, so we expect all the tasks to finish eventually, so no need for extra timeout :)

```suggestion
    |> Task.await_many(:infinity)
```

mruoss replied:
yup right.

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+          class="flex gap-2 mt-4 items-center"
+          autocomplete="off"
+          spellcheck="false"
+          errors={["asdf"]}
```
```suggestion
```

jonatanklosko commented on test/livebook_web/live/session_live_test.exs:
```diff
@@ -1171,6 +1171,199 @@ defmodule LivebookWeb.SessionLiveTest do
+      |> element(~s{button[phx-click="new_pvc"]})
+      |> render_click()
+
+      assert view
+             |> element(~s{form[phx-submit="create_pvc"] input[name="pvc[name]"]})
+             |> has_element?()
+
+      assert view
+             |> element(
+               ~s{form[phx-submit="create_pvc"] input[name="pvc[size_gb]"][type="number"]}
+             )
+             |> has_element?()
+
+      assert view
+             |> element(
+               ~s{form[phx-submit="create_pvc"] select[name="pvc[access_mode]"] option[value="ReadWriteOnce"][selected]}
+             )
+             |> has_element?()
+
+      assert view
+             |> element(
+               ~s{form[phx-submit="create_pvc"] select[name="pvc[storage_class]"] option[value="first-storage-class"]}
+             )
+             |> has_element?()
+
+      assert view
+             |> element(
+               ~s{form[phx-submit="create_pvc"] select[name="pvc[storage_class]"] option[value="second-storage-class"]}
+             )
+             |> has_element?()
+
+      assert view
+             |> element(~s{form[phx-submit="create_pvc"] button[type="submit"][disabled]})
+             |> has_element?()
```
Instead of checking for each element, it should be enough to check if the form is there, no? But then, we don't need to do that either, because `render_change` below would fail if the form was missing.

mruoss replied:
true in some cases. I still want to test that the storage_class options are there as they are loaded dynamically. And I'd like to test the logic of disabling the submit button. i.e. disabled before and enabled after validation. wdyt?

mruoss replied:
I'll also add a cancel button intermezzo

jonatanklosko commented on test/livebook_web/live/session_live_test.exs:
```diff
@@ -1171,6 +1171,199 @@ defmodule LivebookWeb.SessionLiveTest do
+        )
+
+      Req.Test.stub(Livebook.Runtime.K8s, Livebook.ClusterStub)
+      Req.Test.stub(Livebook.Runtime.K8s.NoPermissionCluster, Livebook.NoPermissionClusterStub)
```
```suggestion
```

mruoss replied:
Hmm... if I remove this, I get:

```
     ** (EXIT from #PID<0.1004.0>) an exception was raised:
         ** (RuntimeError) cannot find mock/stub Livebook.Runtime.K8s.NoPermissionCluster in process #PID<0.1011.0>
             (req 0.5.2) lib/req/test.ex:388: Req.Test.__fetch_plug__/1
             (req 0.5.2) lib/req/test.ex:621: Req.Test.call/2
             (req 0.5.2) lib/req/steps.ex:894: Req.Steps.run_plug/1
             (req 0.5.2) lib/req/request.ex:1101: Req.Request.run_request/1
             (req 0.5.2) lib/req/request.ex:1045: Req.Request.run/1
             (livebook 0.15.0-dev) lib/livebook/k8s/auth.ex:51: Livebook.K8s.Auth.create_self_subject_access_review/2
             (elixir 1.16.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
             (elixir 1.16.2) lib/task/supervised.ex:36: Task.Supervised.reply/4
```

jonatanklosko replied:
Interesting, at a glance I though we don't interact with the no-permission cluster in this test.

jonatanklosko replied:
Or is the no-permission picked as the default? Then I think the error would be solved by moving `set_context` from mount to update as per another comment.

mruoss replied:
oooh that's it! Nice catch!

jonatanklosko commented on test/livebook_web/live/session_live_test.exs:
```diff
@@ -1171,6 +1171,199 @@ defmodule LivebookWeb.SessionLiveTest do
+
+      Session.subscribe(session.id)
+
+      Req.Test.stub(Livebook.Runtime.K8s, Livebook.ClusterStub)
+      Req.Test.stub(Livebook.Runtime.K8s.NoPermissionCluster, Livebook.NoPermissionClusterStub)
```
`Livebook.Runtime.K8s` here has nothing to do with the actual module, so I would change the naming for clarity:

```suggestion
      Req.Test.stub(:k8s_cluster, Livebook.K8sClusterStub)
      Req.Test.stub(:k8s_no_permission_cluster, Livebook.K8sNoPermissionClusterStub)
```

Sidenote: just checking, there is no way to detect the context from the request?

mruoss replied:
Not sure I get what you mean... I'd like to test the context switcher. Therefore I need to have 2 contexts in my Kubeconfig (see test.exs and `Kubereq.Kubeconfig.Stub`).

jonatanklosko replied:
I was wondering if we can have a single plug implementation that we would stub for both contexts. But this would only work if we can get context from request, so we can deny permission if the request comes for "no-permission".

mruoss replied:
yeah I see. gotta think about this. :)

I guess if `Kubereq.Kubeconfig.Plug` would set the server_url to something like `"https://#{context_name}"` [here](https://github.com/mruoss/kubereq/blob/v0.1.7/lib/kubereq/kubeconfig/stub.ex#L123), we could decide based on that. But I'd have to patch `kubereq` first.

mruoss replied:
Yup, this worked. What do you think of it now?

jonatanklosko replied:
Looks good!

jonatanklosko commented:
@mruoss thank you! I write the changelog when releasing, so no need to worry about that. Great that you enjoyed working on this and the timing was great, so no worries :)

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+    req = socket.assigns.reqs.pvc
+
+    socket =
+      case Kubereq.delete(req, namespace, name) do
```
Let's make this async also :)

mruoss replied:
yup. And creation too, I guess.

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+       docker_tag: hd(Livebook.Config.docker_images()).tag,
+       pod_template: %{template: Pod.default_pod_template(), status: :valid, message: nil}
+     )
+     |> set_context(kubeconfig.current_context)}
```
In case a k8s runtime is already configured with a different context, we will try to load it immediately in `update`. So I would move this call to `update`, I think the default `cond` clause?

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+    manifest = PVC.manifest(pvc, namespace)
+    req = socket.assigns.reqs.pvc
+
+    case Kubereq.create(req, manifest) do
```
Let's make this async also :) See the Fly component where we use `@volume_action.inflight` to show "Creating...", etc. Basically we want to avoid blocking LV on the request.

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+        end),
+        Task.async(fn -> Kubereq.list(reqs.namespaces, nil) end)
+      ]
+      |> Task.await_many(30_000)
```
```suggestion
      |> Task.await_many(:infinity)
```

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+    end
+  end
+
+  defp storage_classes(%{rbac_permissions: %{list_sc: false}}), do: []
```
```suggestion
  defp storage_classes(%{rbac: %{permissions: %{list_sc: false}}}), do: []
```

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+    end
+  end
+
+  defp pvc_options(%{assigns: %{rbac_permissions: %{list_pvc: false}}} = socket) do
```
```suggestion
  defp pvc_options(%{assigns: %{rbac: %{permissions: %{list_pvc: false}{}} = socket) do
```

mruoss replied:
oops

jonatanklosko commented on lib/livebook_web/live/session_live/k8s_runtime_component.ex:
```diff
@@ -0,0 +1,952 @@
+          |> assign(:cluster_check, %{status: :ok, error: nil})
+
+        %Req.Response{status: _other} ->
+          # cannot list namespaces
```
Should we show an error?

mruoss replied:
It is technically not necessary to list namespaces as long as you have permission to create a pod in your target namespace. So what I do in the UI in this case: Show a text input field instead of a select for the target namespace.

jonatanklosko replied:
Sounds good!

jonatanklosko reviewed (approved):
Massive work! Looks good to me, I plan to merge it this week, make another pass to share some code with the Fly runtime, and then we can make a new release :)

jonatanklosko commented on lib/livebook/runtime/k8s.ex:
```diff
@@ -0,0 +1,593 @@
+          {^port, {:data, "Forwarding from " <> _}} ->
+            :ok
+
+          {^port, {:data, "Error: " <> error}} ->
+            {:error, "failed to port-forward. Error: #{String.trim(error)}"}
```
@mruoss please double check. I run:

```
$ kubectl port-forward pod/mypod 5000 6000
Error from server (NotFound): pods "mypod" not found
```

and the error has no `:`, so I think we want this:

```suggestion
          {^port, {:data, "Error " <> _ = message}} ->
            {:error, "failed to port-forward. #{String.trim(message)}"}
```

mruoss replied:
oh wow, what a catch! Yes you're absolutely right. I guess I didn't test this path.
